File: descripe/imapp_modisl2/src/leocat/src/MOD_PRAlg29/get_pn_thresholds.c
/*
!C ********************************************************************
!Description:

  Function get_pn_thresholds.c

  Routine for setting 3.9 minus 11 um cloud test thresholds for use
  in polar night conditions.
 
!Input parameters:
  bt_11         11 um brightness temperature for current pixel
  bt_bnds       11 um brightness temperature bounds for current test
  th_low        Test thresholds for 'bt_11' lt 'bt_bnds(1)'.
  th_mid1       Test thresholds for 'bt_11' between 'bt_bnds(1)' and
                     'bt_bnds(2)'.
  th_mid2       Test thresholds for 'bt_11' between 'bt_bnds(2)' and
                     'bt_bnds(3)'.
  th_mid3       Test thresholds for 'bt_11' between 'bt_bnds(3)' and
                     'bt_bnds(4)'.
  th_hi         Test thresholds for 'bt_11' gt 'bt_bnds(4)'.
 
!Output Parameters:
  locut         Zero confidence value of clear sky confidence interval
  hicut         1.0 confidence value of clear sky confidence interval
  midpt         0.5 confidence value of clear sky confidence interval
  power         Power of clear sky confidence curve

!Revision History:
  Converted to C.
  R. Frey           10/2007


!Team-unique Header:

!References and Credits:

!END ******************************************************************/

/* Includes */

#include <stdio.h>
#include "thresholds.h"

/**********************************************************************/


void get_pn_thresholds(float bt_11, float *bt_bnds, float *th_low,
                    float *th_mid1, float *th_mid2, float *th_mid3, 
                    float *th_hi, float *locut, float *hicut, float *midpt,
                    float *power)


/**********************************************************************/

{

/*  Declarations */

    float lo_tmp;
    float hi_tmp;
    float lo_tmp_thr;
    float hi_tmp_thr;
    float conf_range;
    float a;
    
/**********************************************************************/

/*  Initializations */

    lo_tmp = 0.0;
    hi_tmp = 0.0;
    lo_tmp_thr = 0.0;
    hi_tmp_thr = 0.0;
    conf_range = 0.0;
    a = 0.0;

/**********************************************************************/

/*  Compute clear sky confidence interval boundaries and mid-point. */

    if(bt_11 < bt_bnds[0]) {

      *(hicut) = th_low[2];
      *(locut) = th_low[0];
      *(midpt) = th_low[1];
      *(power) = th_low[3];

    }

    else if(bt_11 > bt_bnds[3]) {

      *(hicut) = th_hi[2];
      *(locut) = th_hi[0];
      *(midpt) = th_hi[1];
      *(power) = th_hi[3];

    }

    else {

      if(bt_bnds[1] == 0.0 && bt_bnds[2] == 0.0) {

        lo_tmp = bt_bnds[0];
        hi_tmp = bt_bnds[3];
        lo_tmp_thr = th_mid1[0];
        hi_tmp_thr = th_mid1[1];
        *(power) = th_mid1[3];
        conf_range = th_mid1[2];

      }

      else if(bt_11 >= bt_bnds[0] && bt_11 < bt_bnds[1]) {

        lo_tmp = bt_bnds[0];
        hi_tmp = bt_bnds[1];
        lo_tmp_thr = th_mid1[0];
        hi_tmp_thr = th_mid1[1];
        *(power) = th_mid1[3];
        conf_range = th_mid1[2];

      }

      else if(bt_11 >= bt_bnds[1] && bt_11 < bt_bnds[2]) {

        lo_tmp = bt_bnds[1];
        hi_tmp = bt_bnds[2];
        lo_tmp_thr = th_mid2[0];
        hi_tmp_thr = th_mid2[1];
        *(power) = th_mid2[3];
        conf_range = th_mid2[2];

      }

      else if(bt_11 >= bt_bnds[2] && bt_11 <= bt_bnds[3]) {

        lo_tmp = bt_bnds[2];
        hi_tmp = bt_bnds[3];
        lo_tmp_thr = th_mid3[0];
        hi_tmp_thr = th_mid3[1];
        *(power) = th_mid3[3];
        conf_range = th_mid3[2];

      }

      a = (bt_11 - lo_tmp) / (hi_tmp - lo_tmp);
      *(midpt) = lo_tmp_thr + (a*(hi_tmp_thr - lo_tmp_thr));
      *(hicut) = *(midpt) - conf_range;
      *(locut) = *(midpt) + conf_range;

    }
/*
    printf("\nget_pn_threshlds: \n");
    printf("%f %f %f %f \n", bt_bnds[0], bt_bnds[1], bt_bnds[2], bt_bnds[3]);
    printf("%f %f %f %f %f %f \n", bt_11, lo_tmp, hi_tmp, lo_tmp_thr, hi_tmp_thr, conf_range);
    printf("%f %f %f %f %f \n", a, *(midpt), *(locut), *(hicut), *(power));
    printf("\n");
*/
    return;

}

File: descripe/imapp_modisl2/src/cloudmask/src/get_pn_thresholds.f
      subroutine get_pn_thresholds(bt_11,bt_bnds,th_low,th_mid1,
     *                             th_mid2,th_mid3,th_hi,locut,hicut,
     *                             midpt,power)

      implicit none
      save

c---------------------------------------------------------------------
C!F77
c
c!Description:
c     Routine for setting cloud test thresholds for 11 minus 3.9 um
c     test in polar night conditions.
c
c!Input parameters:
c bt_11         11 um brightness temperature for current pixel
c bt_bnds       11 um brightness temperature bounds for current test
c th_low        Test thresholds for 'bt_11' lt 'bt_bnds(1)'.
c th_mid1       Test thresholds for 'bt_11' between 'bt_bnds(1)' and 
c                    'bt_bnds(2)'.
c th_mid2       Test thresholds for 'bt_11' between 'bt_bnds(2)' and
c                    'bt_bnds(3)'.
c th_mid3       Test thresholds for 'bt_11' between 'bt_bnds(3)' and
c                    'bt_bnds(4)'.
c th_hi         Test thresholds for 'bt_11' gt 'bt_bnds(4)'.
c
c!Output Parameters:
c locut         Zero confidence value of clear sky confidence interval
c hicut         1.0 confidence value of clear sky confidence interval
c midpt         0.5 confidence value of clear sky confidence interval
c power         Power of clear sky confidence curve
c
c!Revision History:
c Original version - 03/07/2002     
c R. Frey            MODIS Science Group at UW-Madison
c Revised          - 10/30/2003
c R. Frey
c 06/04 Collection 5  R. Frey
c Added 'th_mid1','th_mid2','th_mid3'.
c
c!Team-unique Header:
c
c!References and Credits:
c See Cloud Mask ATBD-MOD-35.
c
c!END
c---------------------------------------------------------------------

c     Array arguments.
      real bt_bnds(4),th_low(4),th_hi(4),th_mid1(4),th_mid2(4),th_mid3(4)

c     Scalar arguments.
      real bt_11,locut,hicut,midpt,power

c     Local scalars.
      real lo_tmp,hi_tmp,lo_tmp_thr,hi_tmp_thr,a,conf_range

c     Compute clear sky confidence interval boundaries and mid-point.

      if(bt_11 .lt. bt_bnds(1)) then

        hicut = th_low(3)
        locut = th_low(1)
        midpt = th_low(2)
        power = th_low(4)

      else if(bt_11 .gt. bt_bnds(4)) then

        hicut = th_hi(3)
        locut = th_hi(1)
        midpt = th_hi(2)
        power = th_hi(4)

      else

        if(bt_bnds(2) .eq. 0.0 .and. bt_bnds(3) .eq. 0.0) then

          lo_tmp = bt_bnds(1)
          hi_tmp = bt_bnds(4)
          lo_tmp_thr = th_mid1(1)
          hi_tmp_thr = th_mid1(2)
          power = th_mid1(4)
          conf_range = th_mid1(3)

        else if(bt_11 .ge. bt_bnds(1) .and. bt_11 .lt. bt_bnds(2)) then

          lo_tmp = bt_bnds(1)
          hi_tmp = bt_bnds(2)
          lo_tmp_thr = th_mid1(1)
          hi_tmp_thr = th_mid1(2)
          power = th_mid1(4)
          conf_range = th_mid1(3)

        else if(bt_11 .ge. bt_bnds(2) .and. bt_11 .lt. bt_bnds(3)) then

          lo_tmp = bt_bnds(2)
          hi_tmp = bt_bnds(3)
          lo_tmp_thr = th_mid2(1)
          hi_tmp_thr = th_mid2(2)
          power = th_mid2(4)
          conf_range = th_mid2(3)

        else if(bt_11 .ge. bt_bnds(3) .and. bt_11 .le. bt_bnds(4)) then

          lo_tmp = bt_bnds(3)
          hi_tmp = bt_bnds(4)
          lo_tmp_thr = th_mid3(1)
          hi_tmp_thr = th_mid3(2)
          power = th_mid3(4)
          conf_range = th_mid3(3)

        end if

        a = (bt_11 - lo_tmp) / (hi_tmp - lo_tmp)
        midpt = lo_tmp_thr + a*(hi_tmp_thr - lo_tmp_thr)
        hicut = midpt - conf_range
        locut = midpt + conf_range
       
      end if

      return
      end
